http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java

```diff
@@ -1240,39 +1244,41 @@ private <I, R extends io.micronaut.http.HttpResponse<?>> Publisher<R> applyFilte
             AtomicReference<MutableHttpRequest<?>> requestWrapper,
             Publisher<R> responsePublisher) {
 
-        if (request instanceof MutableHttpRequest) {
-            ((MutableHttpRequest<I>) request).uri(requestURI);
+        if (!(request instanceof MutableHttpRequest)) {
+            return responsePublisher;
+        }
 
-            List<HttpClientFilter> filters =
-                    filterResolver.resolveFilters(request, clientFilterEntries);
-            if (parentRequest != null) {
-                filters.add(new ClientServerContextFilter(parentRequest));
-            }
+        ((MutableHttpRequest<I>) request).uri(requestURI);
 
-            OrderUtil.reverseSort(filters);
-            Publisher<R> finalResponsePublisher = responsePublisher;
-            filters.add((req, chain) -> finalResponsePublisher);
+        List<GenericHttpFilter> filters =
+                filterResolver.resolveFilters(request, clientFilterEntries);
+        if (parentRequest != null) {
+            // todo: migrate to new filter
+            filters.add(
+                new GenericHttpFilter.AroundLegacy(new ClientServerContextFilter(parentRequest),
+                new FilterOrder.Fixed(Ordered.HIGHEST_PRECEDENCE)));
+        }
 
-            ClientFilterChain filterChain = buildChain(requestWrapper, filters);
-            if (parentRequest != null) {
-                responsePublisher = ServerRequestContext.with(parentRequest, (Supplier<Publisher<R>>) () -> {
-                    try {
-                        return Flux.from((Publisher<R>) filters.get(0).doFilter(request, filterChain))
-                                .contextWrite(ctx -> ctx.put(ServerRequestContext.KEY, parentRequest));
-                    } catch (Throwable t) {
-                        return Flux.error(t);
-                    }
-                });
-            } else {
-                try {
-                    responsePublisher = (Publisher<R>) filters.get(0).doFilter(request, filterChain);
-                } catch (Throwable t) {
-                    responsePublisher = Flux.error(t);
+        FilterRunner.sortReverse(filters);
+        filters.add(new GenericHttpFilter.TerminalReactive(responsePublisher));
+
+        FilterRunner runner = new FilterRunner(filters);
+        Mono<R> responseMono = Mono.deferContextual(ctx -> {
```

Comment:
We should avoid extra storing and propagating the Reactor context, it should be done automatically like before. There can be cases where we might want to detect if the route is already reactive and don't use a completable stage but some reactive one to preserve the context.

Reply:
It would be nice to avoid anything Reactor/Reactive specific, like deferred... Essentially we might want to extract all the reactive code at some point.
If it's needed, we might need to introduce `ExecutionFlow#defered` and `ExecutionFlow#async` which will resolve to reactive/completablefuture based if the current flow is already reactive or not.

Reply:
since filters dont have to use reactive anymore, the only way to propagate the context between filters that do, is to store it in the filter runner

Reply:
i saw that you tried to do context propagation in ExecutionFlow, but imo the model for context propagation is limited. we should not have to wait for the caller to "subscribe" to the ExecutionFlow on the off chance that the caller will write context. The possibility to run the context-dependent code immediately should be left open (and in fact it's what's happening in FilterRunner right now).

This requires passing in the context *before* FilterRunner.run is called, not in some contextWrite that is done after

Reply:
It is limited and might be something to be removed, but I don't see a valid use case where the Reactor context needs to be detached from the actual reactive flow and attached somewhere else.

Reply:
Ehh not sure if I like this blocking stuff... In this case, the context should be somehow propagated in `FilterContinuation`, if it's even possible, I don't think Reactor/Reactive mix well with blocking operations.

Reply:
i just think that filters should be allowed not to use reactive. especially if we want first-class loom support in the long run. by putting the context in a field that is updated throughout the filter runner, the reactive context simply "skips over" filters that dont use it.